### PR TITLE
Web reset + file uploads

### PR DESCRIPTION
**Pull request by joshling1919**
Moving updates from https://github.com/openoakland/woeip/pull/192 to a new branch since it was easier to resolve conflicts this way.

---
name: File Uploads
about: Request merging changes into the code base
title: ''
labels: ''
reviewers: ''

---

## Checklist
- [x ] Add description
- [x ] Reference open issue pull request addresses
- [ ] Pass functional tests
  - complete on the local machine with `make local.shell` `make test`
- [x ] Request code review
  - Please allow **36 hours** from opening a pull request before merging a pull request- even if it has already recieved an approving review.
- [ ] Address comments on code and resolve requested changes
- [ ] Merge own code

## Description
Issue: #129 #184

*Brief description of solution*

image

image

image

image

### CODE REVIEW

File: api/woeip/settings.py [hunk 1]
```diff
@@ -67,6 +68,12 @@
 
 ROOT_URLCONF = "woeip.urls"
 
+CORS_ORIGIN_ALLOW_ALL = False
+
+CORS_ORIGIN_WHITELIST = (
+    'http://lvh.me',
```
Comment (TangoYankee) [inline api/woeip/settings.py]: Will we need to change the whitelist and should it be an open issue?

File: web/src/components/App/index.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import React from 'react'
+import { Route, Switch } from 'react-router-dom'
+import { ThemeProvider } from 'emotion-theming'
+import { theme } from 'theme'
```
Comment (TangoYankee) [inline web/src/components/App/index.tsx:4]: Using just the module name works in running the application. However, the tests are not able to follow the paths. This applies not just to `theme` but all the custom module imports. I talk about it more in issue #196.

File: web/package.json [hunk 3]
```diff
@@ -25,18 +32,17 @@
...
-    "tslint-react": "^4.0.0",
-    "typescript": "^3.5.2"
+    "tslint-plugin-prettier": "^2.3.0",
+    "tslint-react": "^4.2.0",
+    "typescript": "^3.8.3"
   },
   "jest": {
     "transform": {
-      "\\.(ts|tsx)$": "<rootDir>/node_modules/ts-jest/preprocessor.js",
+      "\\.(ts|tsx)$": "ts-jest",
```
Comment (TangoYankee) [inline web/package.json]: You beat me to the change in PR #197

File: web/src/components/Upload/index.tsx [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+  const [files, setFiles] = useState<Array<FileWithPath>>([])
+  const [success, setSuccess] = useState<boolean>(false)
+  const { getRootProps, getInputProps } = useDropzone({
+    onDrop: (acceptedFiles: Array<FileWithPath>) => {
+      setFiles([...files, ...acceptedFiles])
+    },
+    multiple: true
+  })
+
+  const upload = e => {
```
Comment (TangoYankee) [inline web/src/components/Upload/index.tsx]: This is flagged in the tests as `Parameter 'e' implicitly has an 'any' type`

File: api/woeip/settings.py [hunk 5]
```diff
@@ -30,7 +30,7 @@
     "django.contrib.gis",
 ]
 
-THIRD_PARTY_APPS = ["django_extensions", "rest_framework", "storages", "drf_yasg"]
+THIRD_PARTY_APPS = ["django_extensions", "rest_framework", "storages", "drf_yasg", "corsheaders"]
```
Comment (TangoYankee) [inline api/woeip/settings.py]: Was this install prompted by a specific error? My understanding is that the [nginx server]( https://github.com/openoakland/woeip/blob/master/nginx-proxy/cors.conf) configures the cors headers. This should be how the maps component is able to make a [get request](https://github.com/openoakland/woeip/blob/994710f174c3412bcb66089452b6ea60bc5cdd87/web/src/components/Map/index.tsx#L49) to the pollutant values without Django CORS Header installed/configured.

Comment (joshling1919) [reply api/woeip/settings.py]: ooh good call! 

so it was throwing this error before I added the cors config:
image

That was also back before I actually got the request to go through, so I had thought that was part of the issue, but it looks like it was actually just a red herring. Didn't realize that the nginx server was taking care of cors stuff for us so we might just need to tweak some config there to handle this error message (even though it doesn't actually cause the request to fail). I'll delete all the cors changes from django, thanks for catching this!

**Comment by joshling1919** [comment]
@TangoYankee and I recently discussed how this PR has started to bloat past its initial scope of setting up a happy path upload of the CSV and Log files. We agreed that it would be best if I wrap this up at a point where users can browse and select or drag and drop files and get some helpful warning messages. After getting to that point, we can discuss with the team how to best split up the rest of this feature.

At this point, uploading the files work correctly, but there are hardcoded `starts_at` and `ends_at` values. I also had to update the `api/woeip/apps/air_quality_serializers.py` file so that the file was not nested under a dict. I tried really hard to make it work with the format that the API previously expected (file nested under dict), but I wasn't able to find a way that would prevent the File object itself from being stringified when it was being nested. As a result, those `serializers.py` tests are now failing.

@TangoYankee, you mentioned that maybe I should just remove that hitting of the endpoint altogether and revert any changes to `serializers.py`. Let me know if that's the path you ultimately prefer me taking, and I'd be happy to do it!

**Comment by theecrit** [comment]
Wow, really looking forward to running through this on Tuesday.

File: nginx-proxy/cors.conf [hunk 6]
```diff
@@ -1,4 +1,4 @@
-add_header 'Access-Control-Allow-Origin' '*';
+add_header 'Access-Control-Allow-Origin' '*' always;
```
Comment (joshling1919) [inline nginx-proxy/cors.conf]: this was needed to handle CORS issue whenever the API returned an error:
https://stackoverflow.com/questions/28160982/how-to-make-nginx-cors-configuration-work-when-server-returns-error

**Comment by TangoYankee** [comment]
> @TangoYankee, you mentioned that maybe I should just remove that hitting of the endpoint altogether and revert any changes to `serializers.py`. Let me know if that's the path you ultimately prefer me taking, and I'd be happy to do it!

I think this would be good to discuss with the team at hack night.

File: api/.gitignore [hunk 7]
```diff
@@ -62,6 +62,7 @@ coverage.xml
 # Django stuff:
 *.log
 !woeip/apps/air_quality/tests/data/gps.log
+!woeip/apps/air_quality/tests/data/random.log
```
Comment (TangoYankee) [inline api/.gitignore]: I think this change should be delayed until another PR

File: api/woeip/apps/air_quality/serializers.py [hunk 8]
```diff
@@ -78,10 +78,8 @@ class Meta:
 
     def create(self, validated_data):
 
-        ## Basic validation that we have the right files
-
+        # Basic validation that we have the right files
```
Comment (TangoYankee) [inline api/woeip/apps/air_quality/serializers.py]: I think changes to this file should be delayed until another PR

File: api/woeip/apps/air_quality/tests/data/random.log [hunk 9]
```diff
@@ -0,0 +1,184 @@
+$ADVER,3080,2.3
```
Comment (TangoYankee) [inline api/woeip/apps/air_quality/tests/data/random.log]: I think this file shouldn't be included until the next PR

File: api/woeip/apps/air_quality/tests/test_api.py [hunk 10]
```diff
@@ -14,6 +14,7 @@
```
Comment (TangoYankee) [inline api/woeip/apps/air_quality/tests/test_api.py]: Changes to this file should wait to the next PR

File: nginx-proxy/cors.conf (hunk 6 above)
Comment (TangoYankee) [reply nginx-proxy/cors.conf]: We can make look into making this part of the next version.

File: web/src/client.tsx [hunk 11]
```diff
@@ -2,7 +2,7 @@ import React from 'react'
 import { hydrate } from 'react-dom'
 import { BrowserRouter } from 'react-router-dom'
 
-import App from './App'
+import App from './components/App'
```
Comment (TangoYankee) [inline web/src/client.tsx]: the `./` shouldn't be necessary

File: web/src/components/App/index.test.tsx [hunk 12]
```diff
@@ -2,7 +2,7 @@ import React from 'react'
 import { render, screen } from '@testing-library/react'
 import '@testing-library/jest-dom/extend-expect'
 
-import App from './App'
+import App from '.'
 import { MemoryRouter } from 'react-router-dom'
 
 describe('<App />', () => {
```
Comment (TangoYankee) [inline web/src/components/App/index.test.tsx:8]: The tests for `heading` and `list` broke when the DOM changed. I replaced those two checks with [    expect(screen.getAllByRole('link')).toHaveLength(4)](https://github.com/openoakland/woeip/blob/7abc804a8bfbc6821a9c355698aeb552c3a6fc44/web/src/components/App/index.test.tsx#L15)

File: web/src/components/App/index.test.tsx [hunk 13]
```diff
@@ -2,7 +2,7 @@ import React from 'react'
 import { render, screen } from '@testing-library/react'
 import '@testing-library/jest-dom/extend-expect'
 
-import App from './App'
+import App from '.'
```
Comment (TangoYankee) [inline web/src/components/App/index.test.tsx]: request to read `components/App`

File: web/src/components/App/index.tsx [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+import Navbar from 'components/Navbar'
+
+import routes from 'routes'
+
+import 'semantic-ui-css/semantic.min.css'
```
Comment (TangoYankee) [inline web/src/components/App/index.tsx:9]: Importing this module propagates two changes for jest.
1. Create a mock style file as in [/__ mocks __/styleMock.js](https://github.com/openoakland/woeip/blob/7abc804a8bfbc6821a9c355698aeb552c3a6fc44/web/__mocks__/styleMock.js#L1)
2. reference said file in package.json moduleNameMapper
   - [`"^.+\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js"`](https://github.com/openoakland/woeip/blob/7abc804a8bfbc6821a9c355698aeb552c3a6fc44/web/package.json#L69)
    - P.S: great call on keeping moduleNameMapper

File: web/src/server.tsx [hunk 15]
```diff
@@ -3,7 +3,7 @@ import React from 'react'
 import { renderToString } from 'react-dom/server'
 import { StaticRouter } from 'react-router-dom'
 
-import App from './App'
+import App from './components/App'
```
Comment (TangoYankee) [inline web/src/server.tsx]: request remove `./`

File: web/src/components/Upload/util.ts [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+      if (file.name.endsWith('.csv')) {
+        csvFile = file
+      }
+
+      if (file.name.endsWith('.log')) {
+        logFile = file
+      }
+    })
+
+    if (!logFile || !csvFile) {
```
Comment (TangoYankee) [inline web/src/components/Upload/util.ts:37]: I like the solution of setting the files as they're found

**Review by TangoYankee — CHANGES_REQUESTED** [review]
These changes should help limit the scope of the PR and also get it in sync with Master

**Review by TangoYankee — APPROVED** [review]
